rfcs/2021-10-14-9811-vrl-performance.md

```diff
@@ -0,0 +1,374 @@
+instructions `[.., Constant, 12, ..]` when evaluated will load the constant
+stored in the `values` `Vec` that is found in position 12 onto the stack.
+
+- values
+
+A list of constant values found in the program. Since the bytecode only
+contains integers any actual values must be stored here. This also allows
+literals to be deduped.
+
+- targets
+
+A list of paths used in the program, similar to `values`.
+
+- stack
+
+The Vm is a stack based Vm. Every expression that is evaluated pushes the
+result on the stack. Every operation pulls the values it uses from the stack.
+
+- ip
+
+The instruction pointer points to the next instruction to evaluate.
+
+With each node of the AST compiled down to just a few bytes and all
+instructions held in contiguous memory evaluation of the program should be able
+to take full advantage of the CPU cache which should result in much faster
+execution.
+
+
+#### Calling functions
+
+Calling functions in the stdlib will be a case of evaluating each parameter
+with the results pushed onto the stack.
+
+Since Vrl allows for named parameters and optional parameters, the compiler
+will need to ensure the bytecode evaluates each parameter in a specific order
+(likely the order declared in the function). Bytecode will need to be emmitted
+for parameters that are not specified in the Vrl script to add a default value
+to the stack.
+
+### Flow analysis
+
+One of the expensive operations is path lookup.
+
+There may be certain situations where this can be improved by analyising the
+flow of data at compile time.
+
+For example, with this code:
+
+```coffee
+.thing[3].thang = 3
+log(.thing[3].thang)
+```
+
+the compiler can work out that the variable written to and read from is the
+same and can thus store that data in an array, any access can be made via a
+quick O(1) lookup using the index to the array.
+
+The flow analysis can become fairly complex.
+
+```coffee
+.message = parse_json!(.message)
+.message[3] = parse_json!(.message[2])
+log(.message[3].thing)
+```
+
+Functions like `parse_json` would need access to this array and hints as to
+what paths should be stored.
+
+A lot more thought needs to go into this before we can consider implementing
+it.
+
+#### Use a Bump Allocator?
+
+We could use a library such as [bumpalo](https://crates.io/crates/bumpalo).
+
+A bump allocator will allocated a significant amount of memory up front. This
+memory will then be used
+
+Use a persistent data structure to avoid the initial clone.
+Create an mutated pool
+
+This is probably not ready yet until this PR for `BTreeMap` lands.
+https://github.com/rust-lang/rust/pull/77438
+
+
+### Optimization
+
+With the code as a single dimension array of Bytecode, it could be possible to
+scan the code for patterns and reorganise the Bytecode so it can run in a more
+optimal way.
+
+A lot more thought and research needs to go into this before we can consider
+implementing these changes.
```

Comment:
That would be fun to experiment with!

Reply:
This is related to another idea I've been mulling over. We currently try to avoid FFI costs in VRL by having it execute directly against an event in memory (via a trait), but still end up doing a decent amount of copying and moving things around since the models aren't an exact match. An alternative approach could be to model access to the event more like IO and have VRL execution itself work in a more "pure" context (i.e. provide it inputs, take resulting mutations as outputs). Without that need for tight data-level integration, it'd be simpler to hand off VRL execution to a general-purpose VM/JIT (e.g. wasm, luajit, etc).

That could also be total nonsense, since my knowledge of current VRL execution is quite poor 😄